Use ``BitOperations`` in couple more places

PaulusParssinen opened the pull request:
_Should_ be files that are only included by `NetCoreAppToolCurrent` projects.
- **UPDATE1**: Made ILVerification library consumed by ILVerify tool target `NetCoreAppToolCurrent` (same as target as ILVerify)
- **UPDATE2**: Never mind, I didn't realize it was shipped as NuGet package.. Reverting..
<details>
<summary>answered question</summary>
I want to use this to ask, whether any contributions to ``src/coreclr/tools/Common/Internal/NativeFormat/`` is welcomed as the files around here are largely untouched from their initial import and could utilize some newer APIs by my quick skim through. Should these be left alone?
</details>

PaulusParssinen committed:
Use BitOperations.Log2 in EETypeBuilderHelpers

PaulusParssinen committed:
Use BitOperations.LeadingZeroCount in NativeFormatWriter

Big thanks to @MichalPetryka for showing me simpler alternative to this

am11 commented on src/coreclr/tools/Common/Internal/NativeFormat/NativeFormatWriter.cs:
```diff
@@ -2013,16 +2014,10 @@ public void Append(uint hashcode, Vertex element)
-                ret++;
-            }
-            return ret;
+            return 1 << (sizeof(uint) * 8) - BitOperations.LeadingZeroCount(x);
```
nit: to clarify operator precedence:
```suggestion
            return 1 << ((sizeof(uint) * 8) - BitOperations.LeadingZeroCount(x));
```

PaulusParssinen committed:
Clarify operator precedence

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

huoyaoyuan commented on src/coreclr/tools/Common/Internal/NativeFormat/NativeFormatWriter.cs:
```diff
@@ -2051,7 +2046,7 @@ private void ComputeLayout()
 
             // Round number of buckets up to the power of two
-            _nBuckets = (uint)(1 << HighestBit(bucketsEstimate));
+            _nBuckets = (uint)HighestBit(bucketsEstimate);
```
Should be better to use `RoundUpToPowerOf2`. The old implementation would round up 2^n to next bigger value.

PaulusParssinen replied:
I noticed this comment yesterday too, will use that instead.. thanks!

PaulusParssinen replied:
Actually, I think the commend here is in accurate that the existing code did not round __up__ to the power of two but only set the highest bit.

e.g.
```csharp
int bucketEstimate = 0b0000_1010;
int highestBitSet = 1 << HighestBit(bucketEstimate); // returns: 0b000_1000
int nextPow2 = BitOperations.RoundUptoPowerOf2(bucketEstimate); // return 0b0001_0000
```

tannergooding replied:
Yeah, at a glance it rather looks like the logic for `HighestBit` is incorrect. It claims to be computing the following, where one would normally interpret "rounded up" as  meaning the actual algorithm is `1 + ceil(log2(x)) for non-zero x`
> Returns 1 + log2(x) rounded up, 0 iff x == 0

What's actually happening, however, is that there is no rounding up and it's simply doing `1 + log2(x) for non-zero x`:
```
Console.WriteLine($"{HighestBit(0)} {BitOperations.Log2(0)}");  // 0 0 -- correct
Console.WriteLine($"{HighestBit(1)} {BitOperations.Log2(1)}");  // 1 0 -- correct
Console.WriteLine($"{HighestBit(2)} {BitOperations.Log2(2)}");  // 2 1 -- correct
Console.WriteLine($"{HighestBit(3)} {BitOperations.Log2(3)}");  // 2 1 -- should be 3, as log2(3) is 1.58496...
Console.WriteLine($"{HighestBit(4)} {BitOperations.Log2(4)}");  // 3 2 -- correct
Console.WriteLine($"{HighestBit(5)} {BitOperations.Log2(5)}");  // 3 2 -- should be 4, as log2(5) is 2.321928...
...
```

MichalStrehovsky commented:
> I want to use this to ask, whether any contributions to ``src/coreclr/tools/Common/Internal/NativeFormat/`` is welcomed as the files around here are largely untouched from their initial import and could utilize some newer APIs by my quick skim through. Should these be left alone?

They were not touched because there were no problems with them and they are not perf hotspots. They can be changed. Note that the files with Gen suffix are generated with the https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/Common/Internal/Metadata/NativeFormat/UpdateNativeFormatSources.cmd script so updating them involves updating the generator and re-running it to generate the sources.

PaulusParssinen commented:
> They were not touched because there were no problems with them and they are not perf hotspots. They can be changed. Note that the files with Gen suffix are generated with the https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/Common/Internal/Metadata/NativeFormat/UpdateNativeFormatSources.cmd script so updating them involves updating the generator and re-running it to generate the sources.

Okay, thanks for the advice! Unrelated to this PR but should those genered files be perhaps updated to follow the ``.g.cs`` pattern?

MichalStrehovsky commented:
> Okay, thanks for the advice! Unrelated to this PR but should those genered files be perhaps updated to follow the `.g.cs` pattern? Or is the contribution bar in these parts: if it works, don't touch? 😄

We're not very consistent with doing this (this is not the only "offline generator") but it might be an improvement to rename them to .g.cs.

PaulusParssinen commented:
I'm unable to build the runtime locally for unknown reasons (#101720) so I'll close this for time being..

PaulusParssinen committed:
Merge branch 'dotnet:main' into use-bitoperations-more

PaulusParssinen committed:
It's doing LZCNT, not Log2..

PaulusParssinen committed:
Merge branch 'main' into use-bitoperations-more

PaulusParssinen committed:
I'm blind

PaulusParssinen committed:
LZCNT -> TZCNT..

- Yep..

PaulusParssinen committed:
Use BitOperations.RotateLeft in NodeFactory.NativeLayout.cs

PaulusParssinen committed:
Add required cast

PaulusParssinen committed:
Replace one more rotate with BitOperations in NativeLayout

PaulusParssinen committed:
Merge branch 'main' into use-bitoperations-more

MichalStrehovsky commented:
/azp run runtime-nativeaot-outerloop

PaulusParssinen commented:
Huh, only asserts in arm32 smoke tests..?

MichalStrehovsky commented:
> `Expression: 'codeManager->IsUnwindable(pvAddress) || runtime->IsConservativeStackReportingEnabled()'`

This is unlikely to be caused by this PR. We just fixed a bug with similar assert on ARM64. @VSadov do you know if this could also have a manifestation on ARM32 (looks like this CI triggered after your ARM64 fix was merged)?

PaulusParssinen committed:
Merge branch 'dotnet:main' into use-bitoperations-more

PaulusParssinen committed:
Use BitOperations.RotateLeft in the NAOT/CoreLib OpenMethodResolver.cs

PaulusParssinen committed:
Use BitOperations.RotateLeft in the NAOT/CoreLib RuntimeFieldHandle.cs

PaulusParssinen committed:
Use BitOperations.RotateLeft in the NAOT/CoreLib RuntimeMethodHandle.cs

PaulusParssinen committed:
Use BitOperations.RotateLeft in the shared TypeHashingAlgorithms.cs

PaulusParssinen committed:
Use nuint overload of BitOperations.RotateLeft in the shared CoreLib CastCache.cs

PaulusParssinen committed:
Make ILVerification target NetCoreAppToolCurrent

The only consumer of this library is ILVerify, which itself is NetCoreAppToolCurrent.

tannergooding commented on src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs:
```diff
@@ -179,26 +180,20 @@ public static IntPtr ResolveMethod(IntPtr resolverPtr, RuntimeTypeHandle thisTyp
-            hash2 = (hash2 + _rotl(hash2, 5)) ^ hashCode4;
+            hash1 = (hash1 + (int)BitOperations.RotateLeft((uint)hash1, 5)) ^ hashCode1;
+            hash2 = (hash2 + (int)BitOperations.RotateLeft((uint)hash2, 5)) ^ hashCode2;
+            hash1 = (hash1 + (int)BitOperations.RotateLeft((uint)hash1, 5)) ^ hashCode3;
+            hash2 = (hash2 + (int)BitOperations.RotateLeft((uint)hash2, 5)) ^ hashCode4;
 
-            hash1 += _rotl(hash1, 8);
-            hash2 += _rotl(hash2, 8);
+            hash1 += (int)BitOperations.RotateLeft((uint)hash1, 8);
+            hash2 += (int)BitOperations.RotateLeft((uint)hash2, 8);
```
These can just be `int.RotateLeft(hash1, 5)` (and so on)

PaulusParssinen replied:
Ah, that's much more cleaner. Thanks!

tannergooding commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/RuntimeFieldHandle.cs:
```diff
@@ -64,7 +59,7 @@ public override int GetHashCode()
 
             int hashcode = declaringType.GetHashCode();
-            return (hashcode + _rotl(hashcode, 13)) ^ fieldName.GetHashCode();
+            return (hashcode + (int)BitOperations.RotateLeft((uint)hashcode, 13)) ^ fieldName.GetHashCode();
```
Same here:
```suggestion
            return (hashcode + int.RotateLeft((hashcode, 13)) ^ fieldName.GetHashCode();
```

tannergooding commented on src/coreclr/tools/Common/Internal/Runtime/EETypeBuilderHelpers.cs:
```diff
@@ -193,15 +194,7 @@ internal static uint ComputeValueTypeFieldPaddingFieldValue(uint padding, uint a
-                alignment >>= 1;
-            }
-            Debug.Assert(alignment == 1);
+            uint alignmentLog2 = (uint)BitOperations.TrailingZeroCount(alignment);
```
```suggestion
            uint alignmentLog2 = uint.TrailingZeroCount(alignment);
```

tannergooding commented on src/coreclr/tools/Common/TypeSystem/Common/TypeHashingAlgorithms.cs:
```diff
@@ -36,35 +37,29 @@ public void Append(string src)
                 if ((_numCharactersHashed & 1) == 1)
                 {
-                    _hash2 = (_hash2 + _rotl(_hash2, 5)) ^ src[0];
+                    _hash2 = (_hash2 + (int)BitOperations.RotateLeft((uint)_hash2, 5)) ^ src[0];
```
Basically any of the `BitOperations` APIs where you're having to insert casts can be avoided by just calling the same API directly on the type. Not going to comment on all of them

PaulusParssinen committed:
Apply PR feedback

PaulusParssinen committed:
Revert TypeHashingAlgorithms.cs

* ILVerification was shipped OOB

PaulusParssinen committed:
Revert "Make ILVerification target NetCoreAppToolCurrent"

This reverts commit 6049c0771d2834c451e3eb59d9ba7cf12bb22569.

PaulusParssinen committed:
Merge branch 'dotnet:main' into use-bitoperations-more

PaulusParssinen committed:
Merge branch 'dotnet:main' into use-bitoperations-more

PaulusParssinen committed:
Merge branch 'dotnet:main' into use-bitoperations-more

PaulusParssinen committed:
Merge branch 'dotnet:main' into use-bitoperations-more

MichalStrehovsky reviewed (approved):
Thank you!
